src/audio/drc/drc.c

```diff
@@ -140,353 +141,206 @@ static int drc_setup(struct drc_comp_data *cd, uint16_t channels, uint32_t rate)
  * End of DRC setup code. Next the standard component methods.
  */
 
-static struct comp_dev *drc_new(const struct comp_driver *drv,
-				const struct comp_ipc_config *config,
-				const void *spec)
+static int drc_init(struct processing_module *mod)
 {
-	struct comp_dev *dev = NULL;
-	struct drc_comp_data *cd = NULL;
-	const struct ipc_config_process *ipc_drc = spec;
-	size_t bs = ipc_drc->size;
+	struct module_data *md = &mod->priv;
+	struct comp_dev *dev = mod->dev;
+	struct module_config *cfg = &md->cfg;
+	struct drc_comp_data *cd;
+	size_t bs = cfg->size;
 	int ret;
 
-	comp_cl_info(&comp_drc, "drc_new()");
+	comp_info(dev, "drc_init()");
 
 	/* Check first before proceeding with dev and cd that coefficients
 	 * blob size is sane.
 	 */
 	if (bs > SOF_DRC_MAX_SIZE) {
-		comp_cl_err(&comp_drc, "drc_new(), error: configuration blob size = %u > %d",
-			    bs, SOF_DRC_MAX_SIZE);
-		return NULL;
+		comp_err(dev, "drc_init(), error: configuration blob size = %u > %d",
+			 bs, SOF_DRC_MAX_SIZE);
+		return -EINVAL;
 	}
 
-	dev = comp_alloc(drv, sizeof(*dev));
-	if (!dev)
-		return NULL;
-	dev->ipc_config = *config;
-
 	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
 	if (!cd)
-		goto fail;
-
-	comp_set_drvdata(dev, cd);
+		return -ENOMEM;
 
+	md->private = cd;
 	cd->drc_func = NULL;
 
 	/* Handler for configuration data */
 	cd->model_handler = comp_data_blob_handler_new(dev);
 	if (!cd->model_handler) {
-		comp_cl_err(&comp_drc, "drc_new(): comp_data_blob_handler_new() failed.");
+		comp_err(dev, "drc_init(): comp_data_blob_handler_new() failed.");
+		ret = -ENOMEM;
 		goto cd_fail;
 	}
 
 	/* Get configuration data and reset DRC state */
-	ret = comp_init_data_blob(cd->model_handler, bs, ipc_drc->data);
+	ret = comp_init_data_blob(cd->model_handler, bs, cfg->data);
 	if (ret < 0) {
-		comp_cl_err(&comp_drc, "drc_new(): comp_init_data_blob() failed.");
+		comp_err(dev, "drc_init(): comp_init_data_blob() failed.");
 		goto cd_fail;
 	}
-	drc_reset_state(&cd->state);
 
-	dev->state = COMP_STATE_READY;
-	return dev;
+	drc_reset_state(&cd->state);
+	mod->simple_copy = true;
+	return 0;
 
 cd_fail:
 	comp_data_blob_handler_free(cd->model_handler);
 	rfree(cd);
-fail:
-	rfree(dev);
-	return NULL;
+	return ret;
```

Comment:
do we have any concerns about the dangling pointer left in `mod` here?

Reply:
Yep, in function module_adapter_new() https://github.com/thesofproject/sof/blob/ead25e82a2ca5d04dccc13b3930f0e9ee95e4c35/src/audio/module_adapter/module_adapter.c#L135 mod and dev are freed if there was error.